File: security/semgrep/artemis-js/command-injection.js
// Test fixture for command-injection.yaml — run with `semgrep --test`, gated by selftest.
// Deliberately vulnerable code: never executed, never shipped, and platform does not scan
// itself. `ruleid:` = a finding is expected on the next line; `ok:` = none is.

import { exec, execSync } from 'node:child_process';
import * as cp from 'node:child_process';

const userInput = process.argv[2];

// Regression: semgrep resolves a bare "child_process" specifier back to the qualified
// child_process.exec() form, but it does NOT resolve the "node:"-prefixed specifier —
// which is the style this org writes. Every call below was silently missed before.
// ruleid: artemis-js-child-process-exec
exec(userInput);
// ruleid: artemis-js-child-process-exec
execSync(userInput);
// ruleid: artemis-js-child-process-exec
cp.exec(userInput);
// ruleid: artemis-js-child-process-exec
cp.execSync(userInput);

// A literal command carries no injection risk.
// ok: artemis-js-child-process-exec
exec('ls -la');
// ok: artemis-js-child-process-exec
cp.exec('ls -la');

// A regex .exec() is not a shell call and must never be flagged.
const pattern = /foo/;
// ok: artemis-js-child-process-exec
pattern.exec(userInput);

// ruleid: artemis-js-eval-nonliteral
eval(userInput);
// ok: artemis-js-eval-nonliteral
eval('1 + 1');

File: security/semgrep/artemis-python/command-injection.py
# Test fixture for command-injection.yaml — run with `semgrep --test`, gated by selftest.
# Deliberately vulnerable scanner test data: never execute or copy into a project.
# `ruleid:` = a finding is expected on the next line; `ok:` = none is.

import subprocess
import sys

user_input = sys.argv[1]

# ruleid: artemis-py-subprocess-shell-true
subprocess.run(user_input, shell=True)
# ruleid: artemis-py-subprocess-shell-true
subprocess.call(user_input, shell=True)
# ruleid: artemis-py-subprocess-shell-true
subprocess.Popen(user_input, shell=True)
# ruleid: artemis-py-subprocess-shell-true
subprocess.check_output(user_input, shell=True)
# ruleid: artemis-py-subprocess-shell-true
subprocess.check_call(user_input, shell=True)

# Interpolating into the command is the actual injection, and must stay caught.
# ruleid: artemis-py-subprocess-shell-true
subprocess.run(f"ls {user_input}", shell=True)

# A hardcoded command has no injection surface. Flagging these was the false-positive
# class that trains people to ignore the rule.
# ok: artemis-py-subprocess-shell-true
subprocess.run("ls -la", shell=True)
# ok: artemis-py-subprocess-shell-true
subprocess.check_output("git rev-parse HEAD", shell=True)

# shell=False (the default) is the safe form and must never be flagged.
# ok: artemis-py-subprocess-shell-true
subprocess.run(["ls", "-la", user_input])

# ruleid: artemis-py-eval-exec-nonliteral
eval(user_input)
# ruleid: artemis-py-eval-exec-nonliteral
exec(user_input)
# ok: artemis-py-eval-exec-nonliteral
eval("1 + 1")
